### PR TITLE
Generate and refine documentation for `lib.customisation`

### PR DESCRIPTION
**Pull request by ShamrockLee**
## Description of changes

This PR adds the documentation of the functions defined inside `lib/customisation.nix` into the Nixpkgs manual.

`lib.customisation` provides functions to customise derivation-related functions, derivations and attribute sets. Such functions that shape the Nixpkgs architecture worth to be documented inside the manual.

Cc: @fricklerhandwerk

### Note

- I'm not quite familiar with `makeScope` and `makeScopeWithSplicing'`. Please help check if their types are documented correctly.
- #234651 updates `makeOverridable` to also support builders that accepts function-based attribute recursion. When that one is merged, the type of `makeOverridable` will need to be changed.

  ```diff
  -makeOverridable :: (AttrSet -> A) -> (AttrSet) -> A
  +makeOverridable :: ((AttrSet | (AttrSet -> AttrSet)) -> A) -> (AttrSet | (AttrSet -> AttrSet)) -> A
  ```

### Future work

So far, the `Type:` and `Example:` blocks have to be located in the bottom of the documentation comment, below the description. Adding the `Example:` tag to example code blocks that weaves together with the description is currently not possible.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: lib/customisation.nix [hunk 1]
```diff
@@ -55,6 +59,11 @@ rec {
      injects `override` attribute which can be used to override arguments of
      the function.
 
+     Please refer to "Nixpkgs Contributors Guide" section
+     "<pkg>.overrideDerivation" to learn about `overrideDerivation` and caveats
```
Comment (fricklerhandwerk) [inline lib/customisation.nix:63]:
```suggestion
     Please refer to  documentation on [`<pkg>.overrideDerivation`](#sec-pkg-overrideDerivation) to learn about `overrideDerivation` and caveats
```

This should render cleanly in the HTML manual.

File: lib/customisation.nix [hunk 2]
```diff
@@ -287,7 +320,25 @@ rec {
...
+         -> AttrSet
+
+       Splice ::
+         { pkgsBuildBuild :: AttrSet
+         , pkgsBuildHost :: AttrSet
+         , pkgsBuildTarget :: AttrSet
+         , pkgsHostHost :: AttrSet
+         , pkgsHostTarget :: AttrSet
+         , pkgsTargetTarget :: AttrSet
+         }
```
Comment (fricklerhandwerk) [inline lib/customisation.nix:340]: Will this render to something that makes sense? To be clear, it's not wrong to do it that way, but if it ends up as a messy pile of characters we may have to reconsider how to present it.

**Review by fricklerhandwerk — APPROVED** [review]
LGTM, but I didn't check the type signatures in detail. We could still fix slip-ups down the line. Exposing the documentation will make it way more likely.

File: lib/customisation.nix (hunk 2 above)
Comment (ShamrockLee) [reply lib/customisation.nix:340]: It renders to a multiple-line code block just like most of the `Example` blocks, at least when viewing the rendered HTML file with Firefox.

This is the first time I use multiple-line `Type` block. I'm not sure if it's the proper representation, but it seems to be more clearly than a super-long single line.

**Comment by ShamrockLee** [comment]
Oops! Let me correct the commit message.

**Comment by ShamrockLee** [comment]
Addressed the above suggestions and checked the rendered documentation.

**Comment by ShamrockLee** [comment]
Fix the type signature of `lib.makeOverridable` (the previous one I wrote is wrong), and add parenthesis in some type signatures to  avoid assumptions about the precedence of the pseudo code type operators `|` and `->`.

How do we represent the [sum type](https://en.wikipedia.org/wiki/Tagged_union) in the type signature documentation? Is it `A | B`, `A + B` or `oneOf(A, B)`?

**Comment by fricklerhandwerk** [comment]
We don't have fixed conventions for type annotations and I'd caution to go slowly about them.

I've observed people literally getting scared away from Nix documentation because it assumes familiarity with Haskell culture or uses its conventions without explicit explanation.

We definitely should start a section where the notation is explained, and until then I'm inclined toward verbose rather than symbolic names. I also think it's fine to experiment a bit at the cost of temporary inconsistencies, and fix them up once the bikeshedding converges into written-down rules.

But let's not worry about it in this PR.

**Review by infinisil — APPROVED** [review]
Looking good to me too!

**Pull request merged**